src/app/server/AclStorage.cpp

```diff
@@ -0,0 +1,496 @@
+            NodeId subject;
+            ReturnErrorOnFailure(Convert(tmp, subject));
+            ReturnErrorOnFailure(mEntry.AddSubject(nullptr, subject));
+        }
+        ReturnErrorOnFailure(iterator.GetStatus());
+    }
+
+    if (!mStagingEntry.targets.IsNull())
+    {
+        auto iterator = mStagingEntry.targets.Value().begin();
+        while (iterator.Next())
+        {
+            Target target;
+            ReturnErrorOnFailure(Convert(iterator.GetValue(), target));
+            ReturnErrorOnFailure(mEntry.AddTarget(nullptr, target));
+        }
+        ReturnErrorOnFailure(iterator.GetStatus());
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR AclStorage::EncodableEntry::EncodeForRead(TLV::TLVWriter & writer, TLV::Tag tag, FabricIndex fabric) const
+{
+    ReturnErrorOnFailure(Stage());
+    ReturnErrorOnFailure(mStagingEntry.EncodeForRead(writer, tag, fabric));
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR AclStorage::EncodableEntry::EncodeForWrite(TLV::TLVWriter & writer, TLV::Tag tag) const
+{
+    ReturnErrorOnFailure(Stage());
+    ReturnErrorOnFailure(mStagingEntry.EncodeForWrite(writer, tag));
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR AclStorage::EncodableEntry::Stage() const
+{
+    ReturnErrorOnFailure(mEntry.GetFabricIndex(mStagingEntry.fabricIndex));
+
+    {
+        Privilege privilege;
+        ReturnErrorOnFailure(mEntry.GetPrivilege(privilege));
+        ReturnErrorOnFailure(Convert(privilege, mStagingEntry.privilege));
+    }
+
+    {
+        AuthMode authMode;
+        ReturnErrorOnFailure(mEntry.GetAuthMode(authMode));
+        ReturnErrorOnFailure(Convert(authMode, mStagingEntry.authMode));
+    }
+
+    {
+        size_t count;
+        ReturnErrorOnFailure(mEntry.GetSubjectCount(count));
+        if (count > 0)
+        {
+            for (size_t i = 0; i < count; ++i)
+            {
+                NodeId subject;
+                ReturnErrorOnFailure(mEntry.GetSubject(i, subject));
+                StagingSubject tmp;
+                ReturnErrorOnFailure(Convert(subject, tmp));
+                mStagingSubjects[i] = tmp.nodeId;
+            }
+            mStagingEntry.subjects.SetNonNull(mStagingSubjects, count);
+        }
+        else
+        {
+            mStagingEntry.subjects.SetNull();
+        }
+    }
+
+    {
+        size_t count;
+        ReturnErrorOnFailure(mEntry.GetTargetCount(count));
+        if (count > 0)
+        {
+            for (size_t i = 0; i < count; ++i)
+            {
+                Target target;
+                ReturnErrorOnFailure(mEntry.GetTarget(i, target));
+                ReturnErrorOnFailure(Convert(target, mStagingTargets[i]));
+            }
+            mStagingEntry.targets.SetNonNull(mStagingTargets, count);
+        }
+        else
+        {
+            mStagingEntry.targets.SetNull();
+        }
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR AclStorage::Init(PersistentStorageDelegate & persistentStorage, const FabricTable & fabricTable)
```

Comment:
Validation of the fabric table and its iteration should probably be done outside, to avoid other maintainers trying to possibly do more, or even own a reference to the fabric table, when it's just used for basic iteration of the list of fabrics.

Suggest adding `AclStorage::InitForFabric(XXX)` 

Also suggest making sure we have an AccessControl call for when a fabric is CREATED to start "empty" in case previous storage errors had left some content that would be unwittingly reused. Can be done in follow-up with a new issue.